akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala

```diff
@@ -119,6 +119,45 @@ case object Lookup {
    * and protocol
    */
   def create(serviceName: String): Lookup = new Lookup(serviceName, None, None)
+
+  private val SrvQuery = """^_(.+?)\._(.+?)\.(.+?)$""".r
+
+  private val DomainName = "^((?!-)[A-Za-z0-9-]{1,63}(?<!-)\\.)+[A-Za-z]{2,6}$".r
+
+  /**
+   * Create a service Lookup from a string with format:
+   * _portName._protocol.serviceName.
+   * (as specified by https://www.ietf.org/rfc/rfc2782.txt)
+   *
+   * If the passed string conforms with this format, a SRV Lookup is returned.
+   * The serviceName part must be a valid domain name.
+   *
+   * The string is parsed and dismembered to build a Lookup as following:
+   * Lookup(serviceName).withPortName(portName).withProtocol(protocol)
+   *
+   * If the string doesn't not conform with the SRV format, a simple A/AAAA Lookup is returned
+   * using the whole string as service name.
+   */
+  def fromString(str: String): Lookup =
+    str match {
+      case SrvQuery(portName, protocol, serviceName) if validDomainName(serviceName) ⇒
+        Lookup(serviceName).withPortName(portName).withProtocol(protocol)
+
+      case _ ⇒ Lookup(str)
+    }
+
+  /**
+   * Returns true if passed string conforms with SRV format. Otherwise returns false.
+   */
+  def isValidSrv(srv: String): Boolean =
+    srv match {
+      case SrvQuery(_, _, serviceName) ⇒ validDomainName(serviceName)
+      case _                           ⇒ false
+    }
+
+  private def validDomainName(name: String): Boolean =
+    DomainName.pattern.asPredicate().test(name)
```

Comment:
Do we need to add pre-validation on the string itself from user input? null, empty...

Reply:
And if we did, does the user have to deal with it vs fail fast?

Reply:
I will check what we can do here to make it nicer. Actually, I don't know how the regex pattern will behave with null` or empty string. Well, empty string is probably a non-match. But `null` I have no idea. I hope it's also a non-match.

Reply:
`"^((?!-)[A-Za-z0-9-]{1,63}(?<!-)\\.)+[A-Za-z]{2,6}$".r.pattern.asPredicate().test(null)` produces a `NullPointerException`. Might indeed be nicer to return 'false', as this is also Java API.

Reply:
👍 Let's never propagate NPEs, thanks for checking

Reply:
The `DomainName` pattern is internal and will never be called for a `null`. We catch it earlier. 

We call `validDomainName` inside `parseSrv` and `isValidSrv`, but only if `SrvQuery` gives a match. 

* `isValidSrv` converts it to false already
* `parseSrv` throws `IllegalArgumentException` for any non-valid input, including `null`

Both cases are covered in the tests. 

About `parseSrv`, I initially thought about returning a `Try`, but changed my mind because of Java API. 

I don't think we should return an `Option` or `Optional`. If an illegal String is passed, we want to inform the user that something was wrong and we will have the same Scala/Java API issue. Option or Optional?

If we don't want to throw an exception, then an `InvalidLookup` is a better option.

Reply:
I believe it's better practice to throw NPE rather than IAE for nulls.

In terms of `parseSrv` you could have a `InvalidLookup` null object, or I think you might be able to use `akka.util.OptionVal[Lookup]` which would be a `null` at runtime and for Java users, and a `OptionVal[Lookup]` for Scala users.

Reply:
I don't know compared to the rest of Akka, but I would prefer we didn't introduce a Scala API that throws exceptions or returns nulls.  Yeah, you _could_ use `Try` or `try and catch` but the point is you shouldn't need to.  I think even returning Java's `Optional` would be an improvement.

Reply:
I think it's ok to throw if the intended usage is with strings that are most often known to be valid.
Regarding consistency with other things in Akka one example that comes to mind is `ActorPath.fromString` and it throws `MalformedURLException`